File: tools/verify.py
"""Cross-check TagForge output against OpenCV, apriltag-imgs, and RuneTag reference files."""
import json, os, sys, urllib.request
import numpy as np
import cv2

os.chdir(os.path.join(os.path.dirname(__file__), '..'))
dump = json.load(open('tools/verify_dump.json'))
fails = []
base_imgs = 'https://raw.githubusercontent.com/AprilRobotics/apriltag-imgs/master/'
W_SIZE = {'april_16h5': 4, 'april_25h9': 5, 'april_36h10': 6, 'april_36h11': 6}
os.makedirs('tools/ref_imgs', exist_ok=True)

def check(name, ok):
    print(('PASS ' if ok else 'FAIL ') + name)
    if not ok:
        fails.append(name)

# ---- 1. ArUco / AprilTag classic bit grids vs cv2 ----
CV_NAMES = {
    '4x4_1000': cv2.aruco.DICT_4X4_1000, '5x5_1000': cv2.aruco.DICT_5X5_1000,
    '6x6_1000': cv2.aruco.DICT_6X6_1000, '7x7_1000': cv2.aruco.DICT_7X7_1000,
    'aruco': cv2.aruco.DICT_ARUCO_ORIGINAL,
    'mip_36h12': getattr(cv2.aruco, 'DICT_ARUCO_MIP_36h12', getattr(cv2.aruco, 'DICT_ARUCO_MIP_36H12', None)),
    'april_16h5': cv2.aruco.DICT_APRILTAG_16h5, 'april_25h9': cv2.aruco.DICT_APRILTAG_25h9,
    'april_36h10': cv2.aruco.DICT_APRILTAG_36h10, 'april_36h11': cv2.aruco.DICT_APRILTAG_36h11,
}
for dict_key, ids in dump['grids'].items():
    cvd = CV_NAMES.get(dict_key)
    if cvd is None:
        print(f'SKIP {dict_key}: not in this cv2 build')
        continue
    d = cv2.aruco.getPredefinedDictionary(cvd)
    n = d.markerSize
    for id_str, bits in ids.items():
        mid = int(id_str)
        img = cv2.aruco.generateImageMarker(d, mid, n + 2)  # 1px per module incl. 1-module border
        inner = (img[1:-1, 1:-1] > 127).astype(int)
        mine = np.array(bits).reshape(n, n)
        if dict_key.startswith('april'):
            # OpenCV stores AprilTag families rotated 180° from AprilRobotics' canonical
            # images; we follow AprilRobotics (checked against apriltag-imgs below).
            check(f'grid {dict_key} #{mid} (cv2, 180°)', np.array_equal(np.rot90(mine, 2), inner))
        else:
            check(f'grid {dict_key} #{mid}', np.array_equal(mine, inner))

# ---- 1b. AprilTag classic vs official apriltag-imgs (canonical orientation) ----
AT2_IMG = {'april_16h5': ('tag16h5', 'tag16_05_%05d.png'), 'april_25h9': ('tag25h9', 'tag25_09_%05d.png'),
           'april_36h10': ('tag36h10', 'tag36_10_%05d.png'), 'april_36h11': ('tag36h11', 'tag36_11_%05d.png')}
for dict_key, (folder, pat) in AT2_IMG.items():
    n = W_SIZE[dict_key]
    for id_str, bits in dump['grids'][dict_key].items():
        mid = int(id_str)
        local = f'tools/ref_imgs/{folder}_{mid}.png'
        if not os.path.exists(local):
            try:
                urllib.request.urlretrieve(base_imgs + f'{folder}/' + pat % mid, local)
            except Exception as e:
                print(f'SKIP {dict_key} #{mid} official img ({e})')
                continue
        ref = cv2.imread(local, cv2.IMREAD_GRAYSCALE)
        inner = (ref[2:2 + n, 2:2 + n] > 127).astype(int)   # strip white quiet + black border
        check(f'official {dict_key} #{mid}', np.array_equal(np.array(bits).reshape(n, n), inner))

# ---- 2. AprilTag 3 rendered pixels vs apriltag-imgs PNGs ----
AT3_IMG = {
    'tagStandard41h12': 'tagStandard41h12/tag41_12_%05d.png',
    'tagCircle49h12': 'tagCircle49h12/tag49_12_%05d.png',
    'tagCustom48h12': 'tagCustom48h12/tag48_12_%05d.png',
}
for fam, pat in AT3_IMG.items():
    for id_str, px in dump['at3'][fam].items():
        mid = int(id_str)
        local = f'tools/ref_imgs/{fam}_{mid}.png'
        if not os.path.exists(local):
            try:
                urllib.request.urlretrieve(base_imgs + pat % mid, local)
            except Exception as e:
                print(f'SKIP {fam} #{mid}: download failed ({e})')
                continue
        ref = cv2.imread(local, cv2.IMREAD_GRAYSCALE)
        mine = np.array(px)
        tw = mine.shape[0]
        if ref.shape[0] != tw:  # some imgs may have extra margin
            print(f'NOTE {fam} ref {ref.shape} vs total_width {tw}')
            off = (ref.shape[0] - tw) // 2
            ref = ref[off:off + tw, off:off + tw]
        refbits = (ref > 127).astype(int)
        check(f'apriltag3 {fam} #{mid}', np.array_equal(refbits, mine))

# ---- 3. RuneTag structural checks (RUNE-129 invariants from coding.cpp) ----
# (RuneTagDrawer's sample tags predate the 2015 cyclic coding — not a valid reference.)
for rid, r in dump['rune'].items():
    if 'error' in r:
        check(f'rune #{rid} generates', False)
        continue
    code, bits = r['code'], r['bits']
    ok = (len(code) == 43 and all(0 <= c <= 6 for c in code)
          and len(bits) == 129
          # every sector holds symbol c+1 in 1..7 -> at least one dot per sector
          and all(any(bits[3 * i + j] for j in range(3)) for i in range(43))
          and 0 <= r['index'] < 117649)
    check(f'rune #{rid} -> canonical #{r["index"]} structure', ok)

# ---- 3b. Boards vs OpenCV, pixel-exact ----
for key, b in dump['charuco'].items():
    mine = np.array(b['img'], np.uint8)
    board = cv2.aruco.CharucoBoard((b['sx'], b['sy']), 32.0, 24.0,
                                   cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_6X6_1000))
    if b['legacy']:
        board.setLegacyPattern(True)
    ref = board.generateImage((b['sx'] * 32, b['sy'] * 32))
    check(f'charuco {b["sx"]}x{b["sy"]} legacy={b["legacy"]}', np.array_equal(ref > 127, mine > 0))

g = dump['gridboard']
mine = np.array(g['img'], np.uint8)
gb = cv2.aruco.GridBoard((g['mx'], g['my']), float(g['markerMm']), float(g['sepMm']),
                         cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_6X6_1000))
ref = gb.generateImage((mine.shape[1], mine.shape[0]))
check('gridboard 4x5', np.array_equal(ref > 127, mine > 0))

# ---- 4. End-to-end: rasterize my ArUco tile bits and detect with cv2 ----
def raster_from_bits(bits, n, scale=20):
    g = np.array(bits).reshape(n, n)
    img = np.zeros((n + 2, n + 2), np.uint8)
    img[1:-1, 1:-1] = (g * 255).astype(np.uint8)
    img = np.pad(img, 2, constant_values=255)   # quiet zone
    return cv2.resize(img, None, fx=scale, fy=scale, interpolation=cv2.INTER_NEAREST)

det = cv2.aruco.ArucoDetector(cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_6X6_1000))
img = raster_from_bits(dump['grids']['6x6_1000']['23'], 6)
corners, ids, _ = det.detectMarkers(img)
check('detect 6x6_1000 #23', ids is not None and list(ids.flatten()) == [23])

det = cv2.aruco.ArucoDetector(cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_APRILTAG_36h11))
img = raster_from_bits(dump['grids']['april_36h11']['23'], 6)
corners, ids, _ = det.detectMarkers(img)
check('detect april_36h11 #23', ids is not None and list(ids.flatten()) == [23])

print('\n%d failures' % len(fails))
sys.exit(1 if fails else 0)

File: tools/verify.mjs
// Node harness: loads the browser JS (window-shimmed) and dumps bit grids / dot patterns
// as JSON for cross-checking against OpenCV + reference data.
import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import path from 'path';
import vm from 'vm';

const root = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const ctx = { window: {}, document: undefined, console };
vm.createContext(ctx);
for (const f of ['data/aruco_dicts.js', 'data/apriltag3.js', 'js/runetag.js', 'js/render.js']) {
  vm.runInContext(readFileSync(path.join(root, f), 'utf8'), ctx, { filename: f });
}
const Render = vm.runInContext('Render', ctx);
const RuneTag = vm.runInContext('RuneTag', ctx);
const W = ctx.window;

const out = {};

// ---- ArUco / AprilTag classic: full bit grids (data area only) ----
out.grids = {};
for (const [dict, ids] of Object.entries({
  '4x4_1000': [0, 1, 23, 999], '5x5_1000': [0, 42], '6x6_1000': [0, 23, 777],
  '7x7_1000': [0, 5], 'aruco': [0, 100], 'mip_36h12': [0, 10],
  'april_16h5': [0, 7, 29], 'april_25h9': [0, 34], 'april_36h10': [0, 2319], 'april_36h11': [0, 23, 586],
})) {
  const d = W.ARUCO_DICTS[dict];
  out.grids[dict] = {};
  for (const id of ids) out.grids[dict][id] = Render.decodeBits(d.markers[id], d.width, d.height);
}

// ---- AprilTag3: full rendered pixel image (total_width incl. border), 1 = white ----
out.at3 = {};
for (const fam of Object.keys(W.APRILTAG3)) {
  out.at3[fam] = {};
  for (const id of [0, 1, 42]) {
    const tile = Render.aprilTag3(fam, id, 0);
    const tw = W.APRILTAG3[fam].total_width;
    // rasterize shape list at 1 unit = 1 px
    const px = Array.from({ length: tw }, () => new Array(tw).fill(0));
    for (const s of tile.shapes) {
      if (s.t !== 'rect') continue;
      const v = s.black ? 0 : 1;
      for (let y = Math.round(s.y); y < Math.round(s.y + s.h); y++)
        for (let x = Math.round(s.x); x < Math.round(s.x + s.w); x++)
          if (y >= 0 && y < tw && x >= 0 && x < tw) px[y][x] = v;
    }
    // black background is implicit paper-white outside first rect; first shape covers all
    out.at3[fam][id] = px;
  }
}

// ---- RuneTag: 129 dot flags for a few ids + canonical index mapping ----
out.rune = {};
for (const id of [1, 6, 100, 12345]) {
  try {
    const { code, index } = RuneTag.generate(id);
    out.rune[id] = { index, code, bits: RuneTag.unpack(code).map(b => b ? 1 : 0) };
  } catch (e) {
    out.rune[id] = { error: e.message };
  }
}

// ---- Boards: rasterize at a known px/mm so OpenCV can compare pixel-for-pixel ----
function raster(tile, pxPerUnit) {
  const w = Math.round(tile.wu * pxPerUnit), h = Math.round(tile.hu * pxPerUnit);
  const img = Array.from({ length: h }, () => new Array(w).fill(1)); // 1 = white paper
  for (const s of tile.shapes) {
    const v = s.black ? 0 : 1;
    if (s.t === 'rect') {
      const x0 = Math.round(s.x * pxPerUnit), y0 = Math.round(s.y * pxPerUnit);
      const x1 = Math.round((s.x + s.w) * pxPerUnit), y1 = Math.round((s.y + s.h) * pxPerUnit);
      for (let y = Math.max(0, y0); y < Math.min(h, y1); y++)
        for (let x = Math.max(0, x0); x < Math.min(w, x1); x++) img[y][x] = v;
    }
  }
  return img;
}

// ChArUco: 32mm squares / 24mm markers keeps modules on integer pixels at 32 px/square,
// so a mismatch means real geometry drift, not resampling noise.
out.charuco = {};
for (const [sx, sy, legacy] of [[4, 6, true], [4, 6, false], [5, 7, true], [5, 7, false], [6, 4, true], [3, 3, true]]) {
  const tile = Render.charucoBoard('6x6_1000', sx, sy, 32, 24, legacy, 0);
  out.charuco[`${sx}x${sy}_${legacy}`] = { img: raster(tile, 32), sx, sy, legacy };
}

// Grid board: units are mm, 32mm markers / 8mm gap, 1 px/mm
out.gridboard = { img: raster(Render.gridBoard('6x6_1000', 4, 5, 32, 8, 0, 0), 1), mx: 4, my: 5, markerMm: 32, sepMm: 8 };

// ---- glyph search: templates and routing, no index needed ----
vm.runInContext(readFileSync(path.join(root, 'js/search.js'), 'utf8'), ctx, { filename: 'search.js' });
const TagSearch = vm.runInContext('TagSearch', ctx);
const ALL_DICTS = ['april_36h11', 'april_36h10', '4x4_1000', '5x5_1000', '6x6_1000', '7x7_1000', 'aruco'];
let glyphFails = 0;
const gcheck = (name, ok, extra = '') => {
  console.log((ok ? 'PASS ' : 'FAIL ') + name + (extra ? ' — ' + extra : ''));
  if (!ok) glyphFails++;
};

// Every glyph is 15 bits of a 3x5 cell and must place inside the grid.
for (const [ch, rows] of Object.entries(TagSearch.FONT3x5)) {
  if (rows.length !== 15) gcheck(`font "${ch}" is 3x5`, false, `${rows.length} cells`);
}
gcheck('font covers A-Z and 0-9', Object.keys(TagSearch.FONT3x5).length === 36,
       `${Object.keys(TagSearch.FONT3x5).length} glyphs`);
gcheck('glyph needs at least 5 modules', TagSearch.glyphTemplate('A', 4) === null);
gcheck('"letter A", "A" and "the letter a" all route to the same glyph',
       TagSearch.asGlyph('letter A') === 'A' && TagSearch.asGlyph('A') === 'A' &&
       TagSearch.asGlyph('the letter a') === 'A');
gcheck('a multi-word concept is not treated as a glyph', TagSearch.asGlyph('a smiling face') === null);

// Known best matches. These are fixed by the dictionaries, so they pin the scoring.
for (const [ch, dict, id, score] of [['L', 'aruco', 64, 0.88], ['U', '5x5_1000', 809, 0.92],
                                     ['N', 'aruco', 1023, 0.92]]) {
  const r = TagSearch.glyph(ch, ALL_DICTS, 1);
  const top = r.results[0];
  gcheck(`best "${ch}" is ${dict} #${id}`,
         top && top.dict === dict && top.id === id && Math.abs(top.score - score) < 0.005,
         top ? `${top.dict} #${top.id} at ${top.score.toFixed(2)}` : 'no result');
}
if (glyphFails) process.exitCode = 1;

writeFileSync(path.join(root, 'tools', 'verify_dump.json'), JSON.stringify(out));
console.log('dumped tools/verify_dump.json');

// quick internal sanity for runetag: aligned code must re-align to itself
const { code, index } = RuneTag.generate(1);
const again = RuneTag.generate(index);
console.log('rune generate(1) -> canonical', index, '| generate(canonical) ->', again.index,
            '| stable:', again.index === index);
